// lodash/mergeWith.js
import { FetchHttpHandler as RequestHandler } from "@aws-sdk/fetch-http-handler";
import { CognitoIdentityClientConfig } from "./CognitoIdentityClient";
export declare const getRuntimeConfig: (
  config: CognitoIdentityClientConfig
) => {
  runtime: string;
  defaultsMode: import("@aws-sdk/types").Provider<
    import("@aws-sdk/smithy-client").ResolvedDefaultsMode
  >;
  bodyLengthChecker: import("@aws-sdk/types").BodyLengthCalculator;
  credentialDefaultProvider: (
    input: any
  ) => import("@aws-sdk/types").Provider<import("@aws-sdk/types").Credentials>;
  defaultUserAgentProvider: import("@aws-sdk/types").Provider<
    import("@aws-sdk/types").UserAgent
  >;
  maxAttempts: number | import("@aws-sdk/types").Provider<number>;
  region: string | import("@aws-sdk/types").Provider<any>;
  requestHandler:
    | (import("@aws-sdk/types").RequestHandler<
        any,
        any,
        import("@aws-sdk/types").HttpHandlerOptions
      > &
        import("@aws-sdk/protocol-http").HttpHandler)
    | RequestHandler;
  retryMode: string | import("@aws-sdk/types").Provider<string>;
  sha256: import("@aws-sdk/types").HashConstructor;
  streamCollector: import("@aws-sdk/types").StreamCollector;
  useDualstackEndpoint: boolean | import("@aws-sdk/types").Provider<boolean>;
