// object-inspect/test/values.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LookupDeveloperIdentityResponseFilterSensitiveLog = exports.LookupDeveloperIdentityInputFilterSensitiveLog = exports.ListTagsForResourceResponseFilterSensitiveLog = exports.ListTagsForResourceInputFilterSensitiveLog = exports.ListIdentityPoolsResponseFilterSensitiveLog = exports.IdentityPoolShortDescriptionFilterSensitiveLog = exports.ListIdentityPoolsInputFilterSensitiveLog = exports.ListIdentitiesResponseFilterSensitiveLog = exports.ListIdentitiesInputFilterSensitiveLog = exports.GetPrincipalTagAttributeMapResponseFilterSensitiveLog = exports.GetPrincipalTagAttributeMapInputFilterSensitiveLog = exports.GetOpenIdTokenForDeveloperIdentityResponseFilterSensitiveLog = exports.GetOpenIdTokenForDeveloperIdentityInputFilterSensitiveLog = exports.GetOpenIdTokenResponseFilterSensitiveLog = exports.GetOpenIdTokenInputFilterSensitiveLog = exports.GetIdentityPoolRolesResponseFilterSensitiveLog = exports.RoleMappingFilterSensitiveLog = exports.RulesConfigurationTypeFilterSensitiveLog = exports.MappingRuleFilterSensitiveLog = exports.GetIdentityPoolRolesInputFilterSensitiveLog = exports.GetIdResponseFilterSensitiveLog = exports.GetIdInputFilterSensitiveLog = exports.GetCredentialsForIdentityResponseFilterSensitiveLog = exports.CredentialsFilterSensitiveLog = exports.GetCredentialsForIdentityInputFilterSensitiveLog = exports.DescribeIdentityPoolInputFilterSensitiveLog = exports.IdentityDescriptionFilterSensitiveLog = exports.DescribeIdentityInputFilterSensitiveLog = exports.DeleteIdentityPoolInputFilterSensitiveLog = exports.DeleteIdentitiesResponseFilterSensitiveLog = exports.UnprocessedIdentityIdFilterSensitiveLog = exports.DeleteIdentitiesInputFilterSensitiveLog = exports.IdentityPoolFilterSensitiveLog = exports.CreateIdentityPoolInputFilterSensitiveLog = exports.CognitoIdentityProviderFilterSensitiveLog = exports.ConcurrentModificationException = exports.DeveloperUserAlreadyRegisteredException = exports.RoleMappingType = exports.MappingRuleMatchType = exports.InvalidIdentityPoolConfigurationException = exports.ExternalServiceException = exports.ResourceNotFoundException = exports.ErrorCode = exports.TooManyRequestsException = exports.ResourceConflictException = exports.NotAuthorizedException = exports.LimitExceededException = exports.InvalidParameterException = exports.InternalErrorException = exports.AmbiguousRoleResolutionType = void 0;
exports.UntagResourceResponseFilterSensitiveLog = exports.UntagResourceInputFilterSensitiveLog = exports.UnlinkIdentityInputFilterSensitiveLog = exports.UnlinkDeveloperIdentityInputFilterSensitiveLog = exports.TagResourceResponseFilterSensitiveLog = exports.TagResourceInputFilterSensitiveLog = exports.SetPrincipalTagAttributeMapResponseFilterSensitiveLog = exports.SetPrincipalTagAttributeMapInputFilterSensitiveLog = exports.SetIdentityPoolRolesInputFilterSensitiveLog = exports.MergeDeveloperIdentitiesResponseFilterSensitiveLog = exports.MergeDeveloperIdentitiesInputFilterSensitiveLog = void 0;
const CognitoIdentityServiceException_1 = require("./CognitoIdentityServiceException");
var AmbiguousRoleResolutionType;
(function (AmbiguousRoleResolutionType) {
    AmbiguousRoleResolutionType["AUTHENTICATED_ROLE"] = "AuthenticatedRole";
    AmbiguousRoleResolutionType["DENY"] = "Deny";
})(AmbiguousRoleResolutionType = exports.AmbiguousRoleResolutionType || (exports.AmbiguousRoleResolutionType = {}));
class InternalErrorException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "InternalErrorException",
            $fault: "server",
            ...opts,
        });
        this.name = "InternalErrorException";
        this.$fault = "server";
        Object.setPrototypeOf(this, InternalErrorException.prototype);
    }
}
exports.InternalErrorException = InternalErrorException;
class InvalidParameterException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "InvalidParameterException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidParameterException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidParameterException.prototype);
    }
}
exports.InvalidParameterException = InvalidParameterException;
class LimitExceededException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "LimitExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "LimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, LimitExceededException.prototype);
    }
}
exports.LimitExceededException = LimitExceededException;
class NotAuthorizedException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "NotAuthorizedException",
            $fault: "client",
            ...opts,
        });
        this.name = "NotAuthorizedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NotAuthorizedException.prototype);
    }
}
exports.NotAuthorizedException = NotAuthorizedException;
class ResourceConflictException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "ResourceConflictException",
            $fault: "client",
            ...opts,
        });
        this.name = "ResourceConflictException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceConflictException.prototype);
    }
}
exports.ResourceConflictException = ResourceConflictException;
class TooManyRequestsException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "TooManyRequestsException",
            $fault: "client",
            ...opts,
        });
        this.name = "TooManyRequestsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TooManyRequestsException.prototype);
    }
}
exports.TooManyRequestsException = TooManyRequestsException;
var ErrorCode;
(function (ErrorCode) {
    ErrorCode["ACCESS_DENIED"] = "AccessDenied";
    ErrorCode["INTERNAL_SERVER_ERROR"] = "InternalServerError";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
class ResourceNotFoundException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        this.name = "ResourceNotFoundException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
    }
}
exports.ResourceNotFoundException = ResourceNotFoundException;
class ExternalServiceException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "ExternalServiceException",
            $fault: "client",
            ..
